memberships/views.py

```diff
@@ -5,30 +5,55 @@
 from django.conf import settings
 from django.contrib.auth import authenticate, login
 from urllib.parse import parse_qs, urlparse
+
+import urllib.request
 import json
 import stripe
 
 from .forms import RegistrationForm
 from .models import Member, Membership
 from .services import StripeGateway
 
+def validate_recaptcha(response):
+    url = 'https://www.google.com/recaptcha/api/siteverify'
+    secret = settings.RECAPTCHA_SECRET_KEY
+    payload = {
+        'secret': secret,
+        'response': response
+    }
+    data = urllib.parse.urlencode(payload).encode("utf-8")
+    request = urllib.request.Request(url, data=data)
+    response = urllib.request.urlopen(request)
+    result = json.loads(response.read().decode())
+    success = result.get('success')
+
+    if (not result.get('success')) or (float(result.get('score')) < 0.5):
+        return 'fail'
+
+    return result
+
+
 
 def register(request):
+
     if request.user.is_authenticated:
         return HttpResponse("cannot register while logged in", status=403)
-
     if not request.method == "POST":
         return render(
             request, "memberships/register.html", {"form": RegistrationForm()}
         )
-
     form = RegistrationForm(request.POST)
+
     if not form.is_valid():
         return render(request, "memberships/register.html", {"form": form})
-
     if not form.cleaned_data["preferred_name"]:
         form.cleaned_data["preferred_name"] = form.cleaned_data["full_name"]
 
+    recaptchaV3_response = request.POST.get('recaptchaV3-response')
+    success = validate_recaptcha(recaptchaV3_response)
+    if success == 'fail':
+        return HttpResponse('Invalid reCAPTCHA response. Please try again.', status=403)
```

Comment:
I actually got a failed recaptcha response once (but without the checkbox) so I know this works, which is great :)

It does make me think that it's possible for normal humans to get this response though, so maybe in a future ticket we can clean this up and redirect back to the register page with a message that explains that google thinks they are a robot?

Reply:
Yeah that makes sense to me, explaining they didn't fill the form out correctly rather than being a robot?

Reply:
Yeah I think that makes sense to me. Or even say that google thinks they are a robot rather than we do.